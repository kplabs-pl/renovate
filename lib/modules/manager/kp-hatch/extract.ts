import type { PackageDependency, PackageFileContent } from "../types";
import { JsonMap, parse } from '@iarna/toml';
import { RANGE_PATTERN } from '@renovatebot/pep440';
import { regEx } from "../../../util/regex";
import { PypiDatasource } from "../../datasource/pypi";
import * as pep440Versioning from '../../versioning/pep440';

export const packagePattern =
  '[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9._-]*[a-zA-Z0-9]';
export const extrasPattern = '(?:\\s*\\[[^\\]]+\\])?';

const rangePattern: string = RANGE_PATTERN;
const specifierPartPattern = `\\s*${rangePattern.replace(
  regEx(/\?<\w+>/g),
  '?:'
)}`;
const specifierPattern = `${specifierPartPattern}(?:\\s*,${specifierPartPattern})*`;
export const dependencyPattern = `(?<package>${packagePattern})(${extrasPattern})(?<specifier>${specifierPattern})`;


function extractBuildSystemRequires(pyproject: JsonMap): PackageDependency[] {
    let buildSystem = pyproject['build-system'] as JsonMap;
    let requires = buildSystem?.requires as string[];

    if(requires == null) return [];

    return requires.map((dep: string) => {
        let r = regEx(dependencyPattern).exec(dep);
        return {
            depName: r?.groups?.['package'],
            currentValue: r?.groups?.['specifier'],
            datasource: PypiDatasource.id,
            versioning: pep440Versioning.id,
            depType: 'devDependencies',
        };
    });
}

function extractProjectDependencies(pyproject: JsonMap): PackageDependency[] {
    let project = pyproject.project as JsonMap;
    let dependencies = project?.dependencies as string[];

    if(dependencies == null) return [];

    return dependencies.map((dep: string) => {
        let r = regEx(dependencyPattern).exec(dep);
        return {
            depName: r?.groups?.['package'],
            currentValue: r?.groups?.['specifier'],
            datasource: PypiDatasource.id,
            versioning: pep440Versioning.id,
            depType: 'dependencies',
        };
    });
}

function extractProjectOptionalDependencies(pyproject: JsonMap): PackageDependency[] {
    let project = pyproject.project as JsonMap;
    let optionalDependencies = project?.['optional-dependencies'] as JsonMap;

    if(optionalDependencies == null) return [];

    let result: PackageDependency[] = [];

    for(let depGroup of Object.keys(optionalDependencies)) {
        let dependencies = optionalDependencies[depGroup] as string[];
        result = [
            ...result,
            ...(dependencies.map((dep: string) => {
                let r = regEx(dependencyPattern).exec(dep);
                return {
                    depName: r?.groups?.['package'],
                    currentValue: r?.groups?.['specifier'],
                    datasource: PypiDatasource.id,
                    versioning: pep440Versioning.id,
                    depType: `${depGroup}Dependencies`,
                };
            }))
        ];
    }

    return result;
}

function extractHatchEnvDependencies(pyproject: JsonMap): PackageDependency[] {
    let tool = pyproject.tool as JsonMap;
    let hatch = tool?.hatch as JsonMap;
    let envs = hatch?.envs as JsonMap;

    if(envs == null) return [];

    let result: PackageDependency[] = [];

    for(let envName of Object.keys(envs)) {
        let dependencies = (envs[envName] as JsonMap).dependencies as string[];
        if(dependencies == null) continue;

        result = [
            ...result,
            ...(dependencies.map((dep: string) => {
                let r = regEx(dependencyPattern).exec(dep);
                return {
                    depName: r?.groups?.['package'],
                    currentValue: r?.groups?.['specifier'],
                    datasource: PypiDatasource.id,
                    versioning: pep440Versioning.id,
                    depType: `${envName}EnvDependencies`,
                };
            }))
        ];
    }

    return result;
}

export async function extractPackageFile(
    content: string,
    packageFile: string
  ): Promise<PackageFileContent | null> {
    const pyproject = parse(content);

    return {
        deps: [
            ...extractBuildSystemRequires(pyproject),
            ...extractProjectDependencies(pyproject),
            ...extractProjectOptionalDependencies(pyproject),
            ...extractHatchEnvDependencies(pyproject),
        ],
    };
}
