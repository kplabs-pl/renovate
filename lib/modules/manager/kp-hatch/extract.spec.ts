import {extractPackageFile} from './extract'

import { Fixtures } from '../../../../test/fixtures';

const pyprojectFull = Fixtures.get('pyproject.full.toml');

describe('modules/manager/kp-hatch/extract', () => {
    describe('extractPackageFile()', () => {
        let filename = '';

        it('empty', async () => {
            const res = await extractPackageFile('', filename);
            expect(res?.deps).toMatchSnapshot();
            expect(res?.deps).toHaveLength(0);
        });

        it('full', async () => {
            const res = await extractPackageFile(pyprojectFull, filename);
            expect(res?.deps).toMatchSnapshot();
            expect(res?.deps).toHaveLength(10);
        });
    });
});