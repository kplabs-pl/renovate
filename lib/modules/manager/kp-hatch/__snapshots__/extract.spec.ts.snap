// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/kp-hatch/extract extractPackageFile() empty 1`] = `[]`;

exports[`modules/manager/kp-hatch/extract extractPackageFile() full 1`] = `
[
  {
    "currentValue": "==1.18.0",
    "datasource": "pypi",
    "depName": "hatchling",
    "depType": "devDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "==0.2.0",
    "datasource": "pypi",
    "depName": "hatch-kp-version",
    "depType": "devDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=1.5",
    "datasource": "pypi",
    "depName": "kp-oryx-utilities",
    "depType": "dependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=1.7",
    "datasource": "pypi",
    "depName": "crcmod",
    "depType": "dependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "==1.4.1",
    "datasource": "pypi",
    "depName": "mypy",
    "depType": "devDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=7.0",
    "datasource": "pypi",
    "depName": "pytest",
    "depType": "devDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=0.5",
    "datasource": "pypi",
    "depName": "inflection",
    "depType": "toolDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=3.2",
    "datasource": "pypi",
    "depName": "django",
    "depType": "toolDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=6.1",
    "datasource": "pypi",
    "depName": "sphinx",
    "depType": "docsEnvDependencies",
    "versioning": "pep440",
  },
  {
    "currentValue": "~=1.2",
    "datasource": "pypi",
    "depName": "flask",
    "depType": "docsEnvDependencies",
    "versioning": "pep440",
  },
]
`;
